File: presentations/python-and-agentic-coding/paper-replication/R/scripts/table2.R
# Replicate AJR (2001) Table 2: OLS Regressions.
#
# Eight specifications regressing income on protection against expropriation
# risk, progressively adding latitude and continent dummies, across the
# whole-world and base samples (cols 7-8 use log output per worker). Mirrors the
# layout on p. 1379. America is the omitted continent dummy.
#
# As in the Python version, the whole-world N is 111 (not the printed 110): AJR's
# maketable2.do notes one observation was mistakenly dropped from the printed
# table, so 111 is the corrected sample (matches the QuantEcon reproduction).
#
# R port of python/scripts/table2.py (reads the same shared data/).

args <- commandArgs(trailingOnly = FALSE)
script_path <- sub("^--file=", "", args[grep("^--file=", args)])
ROOT <- dirname(dirname(normalizePath(script_path)))  # paper-replication/R/
DATA <- file.path(dirname(ROOT), "data")               # shared data at repl root

df <- foreign::read.dta(file.path(DATA, "maketable2.dta"))

# label, outcome, regressors, base_sample_only
SPECS <- list(
  list("(1)", "logpgp95", c("avexpr"), FALSE),
  list("(2)", "logpgp95", c("avexpr"), TRUE),
  list("(3)", "logpgp95", c("avexpr", "lat_abst"), FALSE),
  list("(4)", "logpgp95", c("avexpr", "lat_abst", "asia", "africa", "other"), FALSE),
  list("(5)", "logpgp95", c("avexpr", "lat_abst"), TRUE),
  list("(6)", "logpgp95", c("avexpr", "lat_abst", "asia", "africa", "other"), TRUE),
  list("(7)", "loghjypl", c("avexpr"), FALSE),
  list("(8)", "loghjypl", c("avexpr"), TRUE)
)

# Rows of the table, in paper order.
TERMS <- list(
  c("avexpr",   "Average protection against expropriation risk, 1985-1995"),
  c("lat_abst", "Latitude"),
  c("asia",     "Asia dummy"),
  c("africa",   "Africa dummy"),
  c("other",    '"Other" continent dummy')
)

cell <- function(fit, term) {
  co <- summary(fit)$coefficients
  if (!term %in% rownames(co)) return("")
  sprintf("%.2f (%.2f)", co[term, "Estimate"], co[term, "Std. Error"])
}

results <- lapply(SPECS, function(s) {
  data <- if (s[[4]]) df[which(df$baseco == 1), ] else df
  formula <- as.formula(paste(s[[2]], "~", paste(s[[3]], collapse = " + ")))
  lm(formula, data = data)
})
labels <- sapply(SPECS, `[[`, 1)

out <- sapply(results, function(fit) {
  cells <- sapply(TERMS, function(t) cell(fit, t[1]))
  c(cells,
    sprintf("%.2f", summary(fit)$r.squared),
    as.character(length(fit$residuals)))
})
out <- as.data.frame(out, check.names = FALSE, stringsAsFactors = FALSE)
colnames(out) <- labels
rownames(out) <- c(sapply(TERMS, `[`, 2), "R-squared", "Number of observations")

# Header note: which outcome / sample each column uses.
outcomes <- sapply(SPECS, function(s) if (s[[2]] == "logpgp95") "logGDPpc95" else "logOutput/worker88")
samples  <- sapply(SPECS, function(s) if (s[[4]]) "base" else "world")

options(width = 300)
report <- c(
  "TABLE 2 — OLS REGRESSIONS",
  "(coefficient; standard error in parentheses)",
  paste("outcome:", paste(sprintf("%s=%s", labels, outcomes), collapse = "  ")),
  paste("sample: ", paste(sprintf("%s=%s", labels, samples), collapse = "  ")),
  "",
  capture.output(print(out))
)
report <- paste0(paste(report, collapse = "\n"), "\n")
cat(report)

dir.create(file.path(ROOT, "tables"), showWarnings = FALSE)
outpath <- file.path(ROOT, "tables", "table2.txt")
writeLines(report, outpath, sep = "")
cat(sprintf("saved %s\n", outpath))

File: presentations/python-and-agentic-coding/paper-replication/python/scripts/table2.py
"""
Replicate AJR (2001) Table 2: OLS Regressions.

Eight specifications regressing income on protection against expropriation risk,
progressively adding latitude and continent dummies, across the whole-world and
base samples (and, in cols 7-8, using log output per worker as the outcome).
Mirrors the layout on p. 1379. America is the omitted continent dummy.

Known discrepancy vs the printed paper (whole-world columns):
  - The published whole-world N is 110; reproductions get 111. AJR's own
    maketable2.do notes that one observation was *mistakenly dropped* from the
    printed Table 2, so 111 is the corrected sample. We do not try to match 110.
  - The whole-world continent-dummy coefficients (col 4) differ from the printed
    values because the public maketable2.dta ships a residual `other` dummy
    (~AUS/MLT/NZL) rather than AJR's internal classification. We use the dummies
    as provided -- this matches the canonical QuantEcon reproduction exactly
    (avexpr 0.39, africa -0.92, R^2 0.72, N 111). Do NOT recode Europe into
    `other`; that is not what standard reproductions do.
  See https://python.quantecon.org/ols.html (reproduces this same dataset).
"""
from pathlib import Path

import pandas as pd
import statsmodels.formula.api as smf

ROOT = Path(__file__).parent.parent  # paper-replication/python/
DATA = ROOT.parent / "data"          # shared data lives at the repl root
df = pd.read_stata(DATA / "maketable2.dta")

# (label, outcome, regressors, base_sample_only)
SPECS = [
    ("(1)", "logpgp95", ["avexpr"], False),
    ("(2)", "logpgp95", ["avexpr"], True),
    ("(3)", "logpgp95", ["avexpr", "lat_abst"], False),
    ("(4)", "logpgp95", ["avexpr", "lat_abst", "asia", "africa", "other"], False),
    ("(5)", "logpgp95", ["avexpr", "lat_abst"], True),
    ("(6)", "logpgp95", ["avexpr", "lat_abst", "asia", "africa", "other"], True),
    ("(7)", "loghjypl", ["avexpr"], False),
    ("(8)", "loghjypl", ["avexpr"], True),
]

# Rows of the table, in paper order.
TERMS = [
    ("avexpr", "Average protection against expropriation risk, 1985-1995"),
    ("lat_abst", "Latitude"),
    ("asia", "Asia dummy"),
    ("africa", "Africa dummy"),
    ("other", '"Other" continent dummy'),
]

results = {}
for label, outcome, regressors, base_only in SPECS:
    data = df[df["baseco"] == 1] if base_only else df
    formula = f"{outcome} ~ " + " + ".join(regressors)
    fit = smf.ols(formula, data=data).fit()
    results[label] = fit


def cell(fit, term):
    if term not in fit.params:
        return ""
    return f"{fit.params[term]:.2f} ({fit.bse[term]:.2f})"


table = {label: [cell(fit, term) for term, _ in TERMS] for label, fit in results.items()}
for label, fit in results.items():
    table[label].append(f"{fit.rsquared:.2f}")
    table[label].append(str(int(fit.nobs)))

index = [lbl for _, lbl in TERMS] + ["R-squared", "Number of observations"]
out = pd.DataFrame(table, index=index)

# Header note: which outcome / sample each column uses.
outcomes = {label: ("logGDPpc95" if oc == "logpgp95" else "logOutput/worker88")
            for label, oc, _, _ in SPECS}
samples = {label: ("base" if b else "world") for label, _, _, b in SPECS}

pd.set_option("display.max_colwidth", None)
pd.set_option("display.width", None)
report = "\n".join([
    "TABLE 2 — OLS REGRESSIONS",
    "(coefficient; standard error in parentheses)",
    "outcome: " + "  ".join(f"{k}={v}" for k, v in outcomes.items()),
    "sample:  " + "  ".join(f"{k}={v}" for k, v in samples.items()),
    "",
    out.to_string(),
]) + "\n"
print(report)

outpath = ROOT / "tables" / "table2.txt"
outpath.write_text(report)
print(f"saved {outpath}")
